File: quick-sort.py


def partition(arr: [], low: int, high: int):
    '''
    Partition the selected part of the list in place.
    That means to select one of the elements,  put all lower to the left, and all higher to the right,
    and the selected one (the pivot) in the middle. In this case we are selecting the rightmost item
    as the pivot.
    :param arr: list to be partitioned
    :param low: low index of list slice (inclusive)
    :param high: high index of list slice (inclusive)
    :return: index of the pivot (the selected element)
    '''
    pivot = arr[high]
    # so the index will have a valid value if we always increment before swapping
    # this is one less than the  midpoint between the lower partition and the higher partition
    partition_index = low - 1
    for j in range(low, high):
        # if the value is greater than pivot, leave the partition index, and the value in place
        if arr[j] < pivot:
            # increment partition index, and move the lower number to the left of it.
            partition_index += 1
            temp = arr[partition_index]
            arr[partition_index] = arr[j]
            arr[j] = temp
    # now move the pivot to the partition index
    partition_index += 1
    temp = arr[high]
    arr[high] = arr[partition_index]
    arr[partition_index] = temp
    # return the updated pivot index
    return partition_index


def quick_sort(arr: [], low: int, high: int):
    '''
    recursive quick sort.
    This is a sort-in-place which means the indices will select which part of the list is being sorted
    :param arr: list to be sorted
    :param low: low index of arr (inclusive)
    :param high: high index of arr (inclusive)
    :return:
    '''

    # termination step
    if low >= high:
        return

    # action step
    pivot_index = partition(arr, low, high)
    quick_sort(arr, low, pivot_index-1)
    quick_sort(arr, pivot_index+1, high)


# let's try it out!


arr = [9, 1, 4, 5, 2, 8, 3, 6, 7]
print("Before quicksort {}".format(arr))
quick_sort(arr, 0, len(arr) - 1)
print("After quicksort {}".format(arr))

File: src/preparations/QuickSort.java
package preparations;

import java.util.Arrays;
import java.util.Stack;

public class QuickSort {

    /**
     * partitions the list in place into low and high slices, relative to the pivot.
     * The pivot is arbitrarily selected as the last element of the list
     * 
     * @param arr  list to be partitioned
     * @param low  low index of slice of list to be partitioned
     * @param high high index of slice of list to be partitioned
     * @return index of the pivot, after partitioning
     */
    private int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int partitionIndex = -1;

        for (int i = 0; i < high; ++i) {
            if (arr[i] < pivot) {
                partitionIndex += 1;
                int temp = arr[partitionIndex];
                arr[partitionIndex] = arr[i];
                arr[i] = temp;
            }
        }
        partitionIndex += 1;
        int temp = arr[partitionIndex];
        arr[partitionIndex] = arr[high];
        arr[high] = temp;

        return partitionIndex;
    }

    /**
     * Sorts the list quickly and recursively
     * 
     * @param arr  list to be sorted
     * @param low  low slice (inclusive)
     * @param high high slice (inclusive)
     */
    private void quickSortRecursive(int[] arr, int low, int high) {
        if (low >= high) {
            return;
        }
        int partitionIndex = partition(arr, low, high);
        this.quickSortRecursive(arr, low, partitionIndex - 1);
        this.quickSortRecursive(arr, partitionIndex + 1, high);
    }

    /**
     * sorts list quickly and iteratively
     * @param arr list to be sorted
     * @param low low index, inclusive
     * @param high high index, inclusive
     */
    private void quickSortIterative(int[] arr, int low, int high) {
    	
    	if (low == high || arr.length == 0) {
    		return;
    	}
    	// entry order: low, pivot-1, pivot+1, high
    	Stack<Integer> stack = new Stack<Integer>();

    	int pivotIndex = this.partition(arr, low, high);
    	if (low < pivotIndex) {
    		stack.push(low);
    		stack.push(pivotIndex-1);
    	}
    	if (high > pivotIndex) {
    		stack.push(pivotIndex+1);
    		stack.push(high);
    	}
    	while (!stack.empty()) {
    		high = stack.pop();
    		low = stack.pop();
    		if (low >= high) {
    			continue;
    		}
    		pivotIndex = partition(arr, low, high);
        	if (low < pivotIndex) {
        		stack.push(low);
        		stack.push(pivotIndex-1);
        	}
        	if (high > pivotIndex) {
        		stack.push(pivotIndex+1);
        		stack.push(high);
        	}
    	}
    }

    /**
     * Runs the quicksort test
     * 
     * @param argv not used
     */
    public static void main(String[] argv) {
        int[] arrRecurse = { 9, 1, 4, 5, 2, 8, 3, 6, 7 };
        int[] arrIter = { 9, 1, 4, 5, 2, 8, 3, 6, 7 };
        System.out.println("Before quicksort: " + Arrays.toString(arrRecurse));
        int low = 0;
        int high = arrRecurse.length - 1;

        QuickSort qs = new QuickSort();

        qs.quickSortRecursive(arrRecurse, low, high);
        System.out.println("After quicksort recursive: " + Arrays.toString(arrRecurse));

        qs.quickSortIterative(arrIter, low, high);
        System.out.println("After quicksort iterative: " + Arrays.toString(arrIter));
    }
}
